pkg/spanfilter/policymatch/attribute.go

```diff
@@ -0,0 +1,140 @@
+type AttributePolicyMatch struct {
+	filters []AttributeFilter
+}
+
+// NewAttributePolicyMatch returns a new AttributePolicyMatch with the given filters. If no filters are given, then the policy matches all spans.
+func NewAttributePolicyMatch(filters []AttributeFilter) *AttributePolicyMatch {
+	return &AttributePolicyMatch{filters: filters}
+}
+
+// Matches returns true if the given span matches the policy.
+func (p *AttributePolicyMatch) Matches(attrs []*commonv1.KeyValue) bool {
+	// If there are no filters, then the span matches.
+	if len(p.filters) == 0 {
+		return true
+	}
+
+	// If there are no attributes, then the span does not match.
+	if len(attrs) == 0 {
+		return false
+	}
+
+	for _, pa := range p.filters {
+		if !matchesAnyFilter(pa, attrs) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func matchesAnyFilter(pa AttributeFilter, attrs []*commonv1.KeyValue) bool {
+	for _, attr := range attrs {
+		// If the attribute key does not match, then it cannot match the policy.
+		if pa.key != attr.Key {
+			continue
+		}
+		switch pa.typ {
+		case StringAttributeFilter:
+			return pa.stringValue == attr.Value.GetStringValue()
+		case Int64AttributeFilter:
+			return pa.int64Value == attr.Value.GetIntValue()
+		case Float64AttributeFilter:
+			return pa.float64Value == attr.Value.GetDoubleValue()
+		case BoolAttributeFilter:
+			return pa.boolValue == attr.Value.GetBoolValue()
+		case RegexAttributeFilter:
+			return pa.regex.MatchString(attr.Value.GetStringValue())
+		}
+	}
+	return false
+}
+
+type AttributeFilterType int
+
+const (
+	StringAttributeFilter AttributeFilterType = iota
+	Int64AttributeFilter
+	Float64AttributeFilter
+	BoolAttributeFilter
+	RegexAttributeFilter
+)
+
+// AttributeFilter is a filter that matches spans based on their attributes.
+type AttributeFilter struct {
+	key          string
+	typ          AttributeFilterType
+	stringValue  string
+	int64Value   int64
+	float64Value float64
+	boolValue    bool
+	regex        *regexp.Regexp
+}
+
+// NewAttributeFilter returns a new AttributeFilter based on the match type.
+func NewAttributeFilter(matchType config.MatchType, key string, value interface{}) (AttributeFilter, error) {
+	if matchType == config.Regex {
+		return NewRegexpAttributeFilter(key, value)
+	}
+	return NewStrictAttributeFilter(key, value), nil
+}
+
+// NewStrictAttributeFilter returns a new AttributeFilter that matches against the given value.
+func NewStrictAttributeFilter(key string, value interface{}) AttributeFilter {
+	attr := AttributeFilter{
+		key: key,
+	}
+
+	switch v := value.(type) {
+	case string:
+		attr.typ = StringAttributeFilter
+		attr.stringValue = v
+	case int:
+		attr.typ = Int64AttributeFilter
+		attr.int64Value = int64(v)
+	case int64:
+		attr.typ = Int64AttributeFilter
+		attr.int64Value = v
+	case float64:
+		attr.typ = Float64AttributeFilter
+		attr.float64Value = v
+	case bool:
+		attr.typ = BoolAttributeFilter
+		attr.boolValue = v
+	}
```

Comment:
i think we should default to returning an error here. that way if someone tries to create a filter for a type we don't support it will be apparent